Add default wave files (#881)
bottlerocket-os/bottlerocket: sources/updater/waves/README.md

# Update waves

As mentioned in the [Bottlerocket update overview](../README.md), OS updates can include "waves" for staggered deployment.
These waves are defined in the `manifest.json`, which lives in the TUF repository.

> Each time an OS update is made available, `manifest.json` is updated with the information pertinant to that update using [updata](../updog) or its related libraries.

Reviewer: ```suggestion
Each time an OS update is made available, `manifest.json` is updated with the information pertinent to that update using [updata](../updog) or its related libraries.
```

Waves may be supplied to `updata` on the command line using arguments, or passed as a TOML-formatted file.

This directory contains a few examples of these update wave files.
Specific details are encapsulated in each file, but they are:

> * `default-waves.toml`: A "normal" (non-emergency) deployment

Reviewer: ```suggestion
* `default-waves.toml`: A "normal" deployment
```

Only putting "non-emergency" on this entry implies that accelerated-waves is an emergency, which isn't exactly true.

* `accelerated-waves.toml`: An accelerated deployment
* `ohno.toml`: An extremely accelerated deployment in case of emergency.

## Understanding waves

Waves include a "seed" and an "end time".

Each Bottlerocket node generates a "seed" for itself which is simply a number between 0-2048 that determines where it falls in the update order.
Nodes that have a seed within the current wave will update.

> All waves include the seeds of the prior wave, so if a node misses its wave for whatever reason, it still update at a later time.

Reviewer 2: Missing an s. updates at a later time.

> The "start time" of a wave signals the end of the current wave and the beginning of the next wave.
> 

Reviewer: ```suggestion
```
This makes it sound like the start time of a wave is the start time of the next wave, which is weird, and recursive without a base case :P

> The initial wave starts at deployment time, and each subsequent wave starts at the end of the prior wave.

Reviewer 2: Sorry this just occurred to me. Do we want a default where the first wave is delayed while double check the prod deployment one last time? In other words, there doesn't seem to be a way to schedule a first wave in the future.

Author: Good point re: the first wave in the future. It is possible to add a "zeroth" wave with an `end_seed` of `0`, and an `end_time` in the future, though nodes with a seed of `0` would still update in this case, albeit a small number.

I'm not quite sure what you mean "double check the prod deployment". I think the only way to test updates is to update, which is why you can set a small number of nodes to update in the first wave.

Reviewer 2: "double check prod deployment" was just meant as an example of why one might want to set the initial wave in the future.  For example, deploying to prod with the first wave in the future.  Then updating an instance using --all or --ignore-waves (forgot exact flags) which would give you time to rollback the repo before any 0 seed nodes start updating.

Anyway the "double check prod" example was just my thought experiment that led me to realize that there might be a flaw with the format since we can not schedule the very first 'in-the-wild' updates in the future with this file format.  That being said, it's not hard to change it in the future and it works for now, so LGTM.

## Writing wave files

Wave files must be [valid TOML](https://github.com/toml-lang/toml).

> Waves must contain two keys, `end_time` and `end_seed`.

Reviewer: ```suggestion
Wave files must be [valid TOML](https://github.com/toml-lang/toml) containing a list of `[[wave]]` entries.

Each wave must contain two keys, `end_time` and `end_seed`.
```

* a valid RFC3339 formatted string OR
* a string like `"7 days"` or `"2 hours"`. Additional details about valid strings can be found [here](../../parse-datetime)

It represents an offset of time starting from when the operator updates the `manifest.json` file, NOT an offset starting at the time `manifest.json` is uploaded to S3. In simple terms, it is "now" plus whatever time period is specified.

> `fleet_percentage` must be an unsigned integer from 0 to 100. It represents the desired total percentage of the fleet to be updated by the time this wave is over.

Reviewer: Is there a reason to allow 0, if these are integers?  0 should mean no-op, so I don't think there's a reason to have it or allow it.

Author: I'll update these words. - the code already enforces 1-100.

> `fleet_percentage` must be an unsigned integer from 0 to 100. It represents the desired total percentage of the fleet to be updated by the time this wave is over.

Reviewer: ```suggestion
`fleet_percentage` must be an unsigned integer from 0 to 100. It represents the desired total percentage of the fleet to be updated by the time this wave is over.
This percentage maps directly to the seed value; it's the percentage of the maximum seed, 2048.
```
